### PR TITLE
expose GC graph via debug handler

### PR DESCRIPTION
**Pull request by deads2k**
Many times when debugging GC problems, it's important to understand the state of the GC graph at a given point in time.  This pull adds the ability to dump that graph in DOT format for later consumption.  It does this by exposing an additional debug handler and allowing any controller init function to produce such a handler that is included under debug.

Sample full output
```
curl http://localhost:10252/debug/controllers/garbagecollector/graph
digraph full {
  // Node definitions.
  0 [
    label="uid=8581a030-9043-11e8-ad4a-54e1ad486dd3
namespace=kube-system
Pod.v1/kube-dns-7b479ccbc6-qz468
"
    group=""
    version="v1"
    kind="Pod"
    namespace="kube-system"
    name="kube-dns-7b479ccbc6-qz468"
    uid="8581a030-9043-11e8-ad4a-54e1ad486dd3"
    missing="false"
    beingDeleted="false"
    deletingDependents="false"
    virtual="false"
  ];
  1 [
    label="uid=822052fc-9043-11e8-ad4a-54e1ad486dd3
namespace=kube-system
Deployment.v1.apps/kube-dns
"
    group="apps"
    version="v1"
    kind="Deployment"
    namespace="kube-system"
    name="kube-dns"
    uid="822052fc-9043-11e8-ad4a-54e1ad486dd3"
    missing="false"
    beingDeleted="false"
    deletingDependents="false"
    virtual="false"
  ];
  2 [
    label="uid=857bd8ac-9043-11e8-ad4a-54e1ad486dd3
namespace=kube-system
ReplicaSet.v1.apps/kube-dns-7b479ccbc6
"
    group="apps"
    version="v1"
    kind="ReplicaSet"
    namespace="kube-system"
    name="kube-dns-7b479ccbc6"
    uid="857bd8ac-9043-11e8-ad4a-54e1ad486dd3"
    missing="false"
    beingDeleted="false"
    deletingDependents="false"
    virtual="false"
  ];

  // Edge definitions.
  0 -> 2;
  2 -> 1;
}
```

You can also select via UID and have all transitive dependencies output:
```
curl http://localhost:10252/debug/controllers/garbagecollector/graph?uid=8581a030-9043-11e8-ad4a-54e1ad486dd3
digraph full {
  // Node definitions.
  0 [
    label="uid=822052fc-9043-11e8-ad4a-54e1ad486dd3
namespace=kube-system
Deployment.v1.apps/kube-dns
"
    group="apps"
    version="v1"
    kind="Deployment"
    namespace="kube-system"
    name="kube-dns"
    uid="822052fc-9043-11e8-ad4a-54e1ad486dd3"
    missing="false"
    beingDeleted="false"
    deletingDependents="false"
    virtual="false"
  ];
  1 [
    label="uid=8581a030-9043-11e8-ad4a-54e1ad486dd3
namespace=kube-system
Pod.v1/kube-dns-7b479ccbc6-qz468
"
    group=""
    version="v1"
    kind="Pod"
    namespace="kube-system"
    name="kube-dns-7b479ccbc6-qz468"
    uid="8581a030-9043-11e8-ad4a-54e1ad486dd3"
    missing="false"
    beingDeleted="false"
    deletingDependents="false"
    virtual="false"
  ];
  2 [
    label="uid=857bd8ac-9043-11e8-ad4a-54e1ad486dd3
namespace=kube-system
ReplicaSet.v1.apps/kube-dns-7b479ccbc6
"
    group="apps"
    version="v1"
    kind="ReplicaSet"
    namespace="kube-system"
    name="kube-dns-7b479ccbc6"
    uid="857bd8ac-9043-11e8-ad4a-54e1ad486dd3"
    missing="false"
    beingDeleted="false"
    deletingDependents="false"
    virtual="false"
  ];

  // Edge definitions.
  1 -> 2;
  2 -> 0;
}
```

And with some sample rendering:
```
curl http://localhost:10252/debug/controllers/garbagecollector/graph  | dot -T svg -o project.svg
```

produces

gc

@kubernetes/sig-api-machinery-pr-reviews 
/assign @caesarxuchao @liggitt 

```release-note
NONE
```

### CODE REVIEW

**Comment by caesarxuchao** [comment]
cc @cheftako who might be interested because this PR adds http handlers to the controller manager.

File: pkg/controller/garbagecollector/dump.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
+package garbagecollector
+
+import (
+	"net/http"
+
```
Comment (caesarxuchao) [inline pkg/controller/garbagecollector/dump.go]: nit: remove the empty lines.

File: pkg/controller/garbagecollector/dump.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+	}
+	deletingDependents := ""
+	if v.deletingDependents {
+		deleting = "(deletingDependents)"
+	}
+	virtual := ""
+	if v.virtual {
+		virtual = "(virtual)"
+	}
+	return fmt.Sprintf(`%s/%s[%s]-%v%s%s%s%s\nmultlinemaybe?`, kind, v.name, v.namespace, v.uid, missing, deleting, deletingDependents, virtual)
```
Comment (caesarxuchao) [inline pkg/controller/garbagecollector/dump.go]: multlinemaybe?

**Review by caesarxuchao — COMMENTED** [review]
The GC changes look good except a few nits.

I'd like us to properly vendor the upstream gonum than depending more on the forked version.

File: pkg/controller/garbagecollector/BUILD [hunk 3]
```diff
@@ -42,14 +43,20 @@ go_library(
         "//staging/src/k8s.io/client-go/tools/cache:go_default_library",
         "//staging/src/k8s.io/client-go/util/retry:go_default_library",
         "//staging/src/k8s.io/client-go/util/workqueue:go_default_library",
+        "//third_party/forked/gonum/graph:go_default_library",
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/controller/garbagecollector/BUILD]: > @liggitt, can we use the upstream gonum/gonum/graph now?Note that the original gonum/graph has been deprecated.

Given how sparse the connections are, the same patch reasoning probably applies.

**Comment by smarterclayton** [comment]
/approve

This sort of debugging capability will become more relevant over time and is consistent with how other `/debug/` endpoints expose information.

**Comment by mfojtik** [comment]
/lgtm

Thanks @deads2k this is really useful.

**Pull request merged**